In [1]:
import pandas as pd
import os

# Data Products

## Extracting Data from Source Folder

In [2]:
# Creating a list with file names in the data_products folder
destination_path = 'data/data_products'
all_files = [file for file in os.listdir(destination_path) if file.endswith('.csv')]

# Create a list of dataframes for each files in the data_products folder
dataframes = []
for i in all_files:
    file = os.path.join(destination_path, i)
    df = pd.read_csv(file)
    dataframes.append(df)

# Concate all dataframes in the 'dataframes' lists
df_products_original = pd.concat(dataframes, ignore_index=True)

In [3]:
# Creating a copy to preserve original dataframes
df_products = df_products_original.copy()

df_products.head()

name main_category  \
0  Lloyd 1.5 Ton 3 Star Inverter Split Ac (5 In 1...    appliances   
1  LG 1.5 Ton 5 Star AI DUAL Inverter Split AC (C...    appliances   
2  LG 1 Ton 4 Star Ai Dual Inverter Split Ac (Cop...    appliances   
3  LG 1.5 Ton 3 Star AI DUAL Inverter Split AC (C...    appliances   
4  Carrier 1.5 Ton 3 Star Inverter Split AC (Copp...    appliances   

       sub_category                                              image  \
0  Air Conditioners  https://m.media-amazon.com/images/I/31UISB90sY...   
1  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
2  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
3  Air Conditioners  https://m.media-amazon.com/images/I/51JFb7FctD...   
4  Air Conditioners  https://m.media-amazon.com/images/I/41lrtqXPiW...   

                                                link ratings no_of_ratings  \
0  https://www.amazon.in/Lloyd-Inverter-Convertib...     4.2         2,255   
1  https://www.amazon.in/LG-Convertible-Anti-Viru...     4.2         2,948   
2  https://www.amazon.in/LG-Inverter-Convertible-...     4.2         1,206   
3  https://www.amazon.in/LG-Convertible-Anti-Viru...     4.0            69   
4  https://www.amazon.in/Carrier-Inverter-Split-C...     4.1           630   

  discount_price actual_price  Unnamed: 0  
0        ₹32,999      ₹58,990         NaN  
1        ₹46,490      ₹75,990         NaN  
2        ₹34,490      ₹61,990         NaN  
3        ₹37,990      ₹68,990         NaN  
4        ₹34,490      ₹67,790         NaN

## Cleaning and Transforming Data

In [4]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1103170 entries, 0 to 1103169
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   name            1103170 non-null  object 
 1   main_category   1103170 non-null  object 
 2   sub_category    1103170 non-null  object 
 3   image           1103170 non-null  object 
 4   link            1103170 non-null  object 
 5   ratings         751582 non-null   object 
 6   no_of_ratings   751582 non-null   object 
 7   discount_price  980844 non-null   object 
 8   actual_price    1067544 non-null  object 
 9   Unnamed: 0      551585 non-null   float64
dtypes: float64(1), object(9)
memory usage: 84.2+ MB


Penyelidikan terhadap data. Untuk sejauh ini kita bisa lihat bahwa terdapat beberapa null value dalam beberapa kolom itu. Kita bisa melihat bahwa untuk kolom ```name, main_category, sub_category, image, link``` seharusnya memiliki tipe data string. Selain itu untuk kolom ```ratings, no_of_ratings, discount_price, actual_price``` yang seharusnya memiliki tipe data interger atau float. Terakhir, terdapat kolom ```Unnamed: 0``` yang perlu kita selidiki.

In [5]:
# Mengubah kolom name, main_category, sub_category, image, dan link menjadi string

df_products['name'] = df_products['name'].astype('string')
df_products['main_category'] = df_products['main_category'].astype('string')
df_products['sub_category'] = df_products['sub_category'].astype('string')
df_products['image'] = df_products['image'].astype('string')
df_products['link'] = df_products['link'].astype('string')

In [6]:
# Untuk melihat bentuk data tpe masing-masing kolom dan jumlah non-null values

df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1103170 entries, 0 to 1103169
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   name            1103170 non-null  string 
 1   main_category   1103170 non-null  string 
 2   sub_category    1103170 non-null  string 
 3   image           1103170 non-null  string 
 4   link            1103170 non-null  string 
 5   ratings         751582 non-null   object 
 6   no_of_ratings   751582 non-null   object 
 7   discount_price  980844 non-null   object 
 8   actual_price    1067544 non-null  object 
 9   Unnamed: 0      551585 non-null   float64
dtypes: float64(1), object(4), string(5)
memory usage: 84.2+ MB


Lalu kita akan lanjut untuk melihat bagaimana cara menghandle numeric data type. Setelah diselidiki ditemukan beberapa non-numeric values dimasing-masing kolom. Oleh karena itu, melainkan menggunakan ```astype()``` method, kita menggunakan ```apply()``` method untuk mengubah values dari string (atau tipe data yang lain) menjadi ```int/float```. Untuk non-numeric kita akan ubah valuenya menjadi ```0``` tapi juga akan tetap melihat kasus unik masing-masing kolom.

In [7]:
# Membuat fungsi untuk mengubah menjadi interger atau float

def getFloat(number):
    try:
        number = float(number)
        return number
    except:
        return None   

In [8]:
df_products['ratings'].unique()

array(['4.2', '4.0', '4.1', '4.3', '3.9', '3.8', '3.5', nan, '4.6', '3.3',
       '3.4', '3.7', '2.9', '5.0', '4.4', '3.6', '2.7', '4.5', '3.0',
       '3.1', '3.2', '4.8', '4.7', '2.5', '1.0', '2.6', '2.8', '2.3',
       '1.7', 'Get', '1.8', '2.4', '2.0', '1.5', '4.9', '1.9', '2.2',
       '1.2', '2.1', '1.4', '1.6', '1.3', 'FREE', '₹2.99', '1.1', '₹70',
       4.5, 3.3, 3.7, 4.2, 4.3, 4.1, 3.8, 4.4, 4.0, 3.5, 3.9, '₹99', 3.4,
       3.6, 3.2, 5.0, 2.6, 4.7, 3.0, 3.1, 4.6, 1.0, 4.9, 2.9, 2.7, 2.2,
       2.5, 4.8, 2.0, 1.7, 1.9, 2.8, 2.4, 1.6, 2.3, 1.8, 1.4, 1.3, 2.1,
       1.5, '₹100', '₹68.99', '₹65'], dtype=object)

In [9]:
df_products['ratings'] = df_products['ratings'].apply(getFloat)

In [10]:
# Menghilangkan ',' agar dapat diubah menjadi interger
df_products['no_of_ratings'] = df_products['no_of_ratings'].str.replace(',', '')

# Untuk bantu check non number valuesnya seperti apa. karena kolom 'no_of_ratings' memiliki unique values sangat banyak
def getNonNumber(number):
    try:
        number = float(number)
        return 0
    except:
        return number

test_norating = df_products['no_of_ratings'].apply(getNonNumber)
test_norating.unique()

# Karena mengetahui bahwa angka non-numeric values merupaka string error yang tidak dimiliki kolom lain, kita dapat mengubahnya menjadi nilai 0

array([0, 'Only 2 left in stock.', 'Only 1 left in stock.',
       'FREE Delivery by Amazon', 'Usually dispatched in 3 to 4 weeks.',
       'Usually dispatched in 5 to 6 days.',
       'Usually dispatched in 4 to 5 days.',
       'Usually dispatched in 6 to 7 days.',
       'Usually dispatched in 7 to 8 days.',
       'Usually dispatched in 11 to 12 days.',
       'Usually dispatched in 4 to 5 weeks.', 'Only 4 left in stock.',
       'Only 3 left in stock.', 'Usually dispatched in 1 to 2 months.',
       'Only 5 left in stock.',
       'This item will be released on August 14 2023.',
       'Usually dispatched in 3 to 5 days.',
       'Usually dispatched in 2 to 3 days.',
       'Usually dispatched in 9 to 10 days.',
       'Usually dispatched in 2 to 3 weeks.',
       'Usually dispatched in 8 to 9 days.'], dtype=object)

In [11]:
df_products['no_of_ratings'] = df_products['no_of_ratings'].apply(getFloat)
df_products['no_of_ratings'] = df_products['no_of_ratings'].astype('Int64')

In [12]:
# Bersihkan string agar dapat diubah
df_products['discount_price'] = df_products['discount_price'].str.replace('₹', '')
df_products['discount_price'] = df_products['discount_price'].str.replace(',', '')

# Karena bekerja dengan currency kita akan menggunaka datatype float
df_products['discount_price'] = df_products['discount_price'].astype(float)

In [13]:
# Bersihkan string agar dapat diubah
df_products['actual_price'] = df_products['actual_price'].str.replace('₹', '')
df_products['actual_price'] = df_products['actual_price'].str.replace(',', '')

# Karena bekerja dengan currency kita akan menggunaka datatype float
df_products['actual_price'] = df_products['actual_price'].astype(float)

In [14]:
# Kita juga membuat kolom baru untuk menyatukan harga yang sudah didiskon dan values yang tidak ada diskonnya.

def getCopyPrice(disc, actual):
    if pd.isnull(disc) and pd.notnull(actual):
        return actual
    else:
        return disc
    

df_products['current_price'] = df_products.apply(lambda x: getCopyPrice(x['discount_price'], x['actual_price']), axis=1)

In [15]:
df_products[df_products['discount_price'].isnull()].head()

name main_category  \
76   LG 1.5 Ton 3 Star Hot & Cold DUAL Inverter Spl...    appliances   
100  Hitachi 1.5 Ton 5 Star Inverter Split AC (Copp...    appliances   
114  Panasonic 1.5 Ton 4 Star Wi-Fi Twin-Cool Inver...    appliances   
126  Hitachi 1.5 Ton 5 Star Window AC (Copper, Dust...    appliances   
137  Portable Air Conditioner,Office Air Conditione...    appliances   

         sub_category                                              image  \
76   Air Conditioners  https://m.media-amazon.com/images/I/51hbo8yQ1E...   
100  Air Conditioners  https://m.media-amazon.com/images/I/41AY1pk5oR...   
114  Air Conditioners  https://m.media-amazon.com/images/I/41Edvsb7Gh...   
126  Air Conditioners  https://m.media-amazon.com/images/I/81Ei0pUgd7...   
137  Air Conditioners  https://m.media-amazon.com/images/I/61zHq-twHL...   

                                                  link  ratings  \
76   https://www.amazon.in/LG-Inverter-Convertible-...      4.0   
100  https://www.amazon.in/Hitachi-Inverter-Copper-...      3.8   
114  https://www.amazon.in/Panasonic-Conditioner-An...      4.5   
126  https://www.amazon.in/Hitachi-Window-Copper-Fi...      2.7   
137  https://www.amazon.in/Portable-Conditioner-Eva...      3.2   

     no_of_ratings  discount_price  actual_price  Unnamed: 0  current_price  
76             265             NaN           NaN         NaN            NaN  
100            748             NaN           NaN         NaN            NaN  
114            195             NaN           NaN         NaN            NaN  
126             10             NaN       49300.0         NaN        49300.0  
137            303             NaN        7250.0         NaN         7250.0

Karena kita sudah mentransformasi masing-masing kolom menjadi data type yang sesuai kita akan menyingkirkan kolom ```Unnamed: 0``` karena dinilai tidak relevan dengan kebutuh analisis data.

In [16]:
df_products = df_products.drop('Unnamed: 0', axis=1)

In [17]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1103170 entries, 0 to 1103169
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   name            1103170 non-null  string 
 1   main_category   1103170 non-null  string 
 2   sub_category    1103170 non-null  string 
 3   image           1103170 non-null  string 
 4   link            1103170 non-null  string 
 5   ratings         739116 non-null   float64
 6   no_of_ratings   739053 non-null   Int64  
 7   discount_price  980844 non-null   float64
 8   actual_price    1067544 non-null  float64
 9   current_price   1067544 non-null  float64
dtypes: Int64(1), float64(4), string(5)
memory usage: 85.2 MB


## Data Demography

Dari kolom yang kita miliki, terdapat 2 kolom kategori yang bisa digunakan untuk melihat pembagian kelompok dengan aggregate functions yaitu `main_category` dan `sub_category`. Lalu kita memiliki kolom values yang bisa mendapatkan insight dari nilai `ratings`, `discount_price` yang akan kita olah untuk melihat besar diskon, dan `actual_price` untuk mendapatkan insight mengenai harga.

In [18]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1103170 entries, 0 to 1103169
Data columns (total 10 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   name            1103170 non-null  string 
 1   main_category   1103170 non-null  string 
 2   sub_category    1103170 non-null  string 
 3   image           1103170 non-null  string 
 4   link            1103170 non-null  string 
 5   ratings         739116 non-null   float64
 6   no_of_ratings   739053 non-null   Int64  
 7   discount_price  980844 non-null   float64
 8   actual_price    1067544 non-null  float64
 9   current_price   1067544 non-null  float64
dtypes: Int64(1), float64(4), string(5)
memory usage: 85.2 MB


In [19]:
# Melihat unique kolom 'main_category'

df_products['main_category'].unique()

<StringArray>
[             'appliances',    'toys & baby products',
             "men's shoes",          'bags & luggage',
         'car & motorbike',                  'stores',
          "men's clothing",          'home & kitchen',
        "women's clothing",             'accessories',
     'tv, audio & cameras',         'beauty & health',
        'sports & fitness',           "kids' fashion",
                   'music', 'grocery & gourmet foods',
            'pet supplies',           "women's shoes",
     'home, kitchen, pets',     'industrial supplies']
Length: 20, dtype: string

In [20]:
# Melihat unique kolom 'sub_category'

df_products['sub_category'].unique()

<StringArray>
[        'Air Conditioners',        'Nursing & Feeding',
             'Casual Shoes',                  'Wallets',
          'Car Accessories',                  'Diapers',
           'Amazon Fashion',         'T-shirts & Polos',
            'Bedroom Linen',             'Toys & Games',
 ...
             'Sports Shoes',          'STEM Toys Store',
        'Strollers & Prams', 'Suitcases & Trolley Bags',
              'Televisions',  'Test, Measure & Inspect',
    'The Designer Boutique',             'Value Bazaar',
             'Western Wear',                     'Yoga']
Length: 112, dtype: string

In [21]:
# Untuk melihat secara general, dataframe berbentuk seperti apa

df_products.describe()

ratings  no_of_ratings  discount_price  actual_price  \
count  739116.000000       739053.0    9.808440e+05  1.067544e+06   
mean        3.832311     840.847039    2.623161e+03  2.311128e+04   
std         0.756101    8651.895695    9.458191e+03  1.355081e+07   
min         1.000000            1.0    8.000000e+00  0.000000e+00   
25%         3.500000            4.0    3.890000e+02  9.900000e+02   
50%         3.900000           20.0    6.790000e+02  1.599000e+03   
75%         4.300000          133.0    1.399000e+03  2.999000e+03   
max         5.000000       589547.0    1.249990e+06  9.900000e+09   

       current_price  
count   1.067544e+06  
mean    2.872724e+03  
std     9.565771e+03  
min     0.000000e+00  
25%     3.990000e+02  
50%     6.990000e+02  
75%     1.549000e+03  
max     1.249990e+06

### Analisis ratings

In [22]:
# Mengagregat value ratings dengan pembagian kelompok kolom 'main_category'

df_products.groupby('main_category')['ratings'].agg(['mean', 'min', 'max', 'count']).sort_values(by='mean', ascending=False)

mean  min  max   count
main_category                                      
grocery & gourmet foods  4.072739  1.0  5.0    6038
toys & baby products     4.030417  1.0  5.0   10514
pet supplies             4.029325  1.0  5.0    2844
beauty & health          4.002085  1.0  5.0   15446
home & kitchen           3.977118  1.0  5.0   26554
music                    3.964741  1.0  5.0    1696
stores                   3.946108  1.0  5.0   56728
bags & luggage           3.939759  1.0  5.0    7314
accessories              3.904224  1.0  5.0  134838
kids' fashion            3.861064  1.0  5.0   12672
appliances               3.840975  1.0  5.0   54712
industrial supplies      3.840392  1.0  5.0    5714
tv, audio & cameras      3.825332  1.0  5.0   99028
sports & fitness         3.805097  1.0  5.0   17656
women's clothing         3.766114  1.0  5.0  128236
men's clothing           3.765749  1.0  5.0   84038
car & motorbike          3.727236  1.0  5.0   11360
women's shoes            3.678979  1.0  5.0    5680
men's shoes              3.638153  1.0  5.0   58008
home, kitchen, pets      3.135000  1.0  5.0      40

In [23]:
# Mengagregat value ratings dengan pembagian kelompok kolom 'main_category' & sub category

df_products.groupby(['main_category', 'sub_category'])['ratings'].agg(['mean', 'min', 'max', 'count']).sort_values(by='mean', ascending=False)

mean  min  max  count
main_category           sub_category                                          
toys & baby products    International Toy Store      4.578947  4.0  4.9     38
beauty & health         Value Bazaar                 4.304545  3.7  4.6    132
grocery & gourmet foods All Grocery & Gourmet Foods  4.176296  2.0  5.0   1890
beauty & health         Health & Personal Care       4.174317  2.4  5.0   2196
toys & baby products    Strollers & Prams            4.127092  1.0  5.0    502
...                                                       ...  ...  ...    ...
tv, audio & cameras     Headphones                   3.503177  1.0  5.0  13786
car & motorbike         Car Electronics              3.497718  1.0  5.0   1490
beauty & health         Personal Care Appliances     3.455056  1.0  5.0    712
tv, audio & cameras     Security Cameras             3.319143  1.0  5.0   8118
home, kitchen, pets     Refurbished & Open Box       3.135000  1.0  5.0     40

[112 rows x 4 columns]

Dengan dua analisis diatas, kita menemukan bahwa 'grocery & gourmet foods' dalam `main_category` memiliki rata2 `rating` terbesar diikuti oleh dengan 'toys & baby products' dan 'beauty & health'.

Melihat lebih detail ke `sub_category`, kita menemukan bahwa hasil tetap konsisten dengan `main_category` sebelumnya.  Dipimpin 'international toy store', 'value bazaar;, dan 'all grocery & gourmet foods' dalam `sub_category`. 

kedua `main_category` dan `sub_category` konsisten dengan urutan terakhir `main_category` dengan 'home, kitchen, pets 'Refurbished & Open Box'.

### Prices

In [24]:
# Mengagregat value current_price dengan pembagian kelompok kolom 'main_category'.

df_products.groupby('main_category')['current_price'].agg(['mean', 'min', 'max', 'count']).sort_values(by='mean', ascending=False)

mean      min         max   count
main_category                                                    
accessories              5615.944755     4.00   499999.00  222342
appliances               5463.668770    20.00   230000.00   62652
music                    4004.035453    50.00    76900.00    2076
stores                   3922.920501     0.00   406009.00   64670
home, kitchen, pets      3643.705882  1199.00     8541.00      34
tv, audio & cameras      3427.445350     0.00  1249990.00  132714
men's shoes              2600.654225    25.00    66649.00  109726
bags & luggage           2317.375915    10.00   140000.00   19964
industrial supplies      2309.831312    35.94   144856.00    8020
sports & fitness         1750.673040    50.00   225250.00   24506
women's shoes            1322.891529    80.00    44986.50   10514
toys & baby products     1247.551715     9.72    49509.00   12196
men's clothing           1028.034104    30.00    40495.95  147556
car & motorbike           989.371209    29.00   119900.00   13974
kids' fashion             987.623114    48.00   406009.00   26412
home & kitchen            911.534746    20.00    60500.00   28946
beauty & health           754.677356     9.00    95000.00   19816
pet supplies              724.201890    37.00    14890.00    3238
women's clothing          649.904524    29.00    49999.00  151624
grocery & gourmet foods   424.183879     8.00    13899.00    6564

In [25]:
# Mengagregat value 'current_prices' dengan menambahkan pembagian kolom 'sub_category'

df_products.groupby(['main_category', 'sub_category'])['current_price'].agg(['mean', 'min', 'max', 'count']).sort_values(by='mean', ascending=False)

mean    min  \
main_category           sub_category                                       
tv, audio & cameras     Televisions                  39849.728827  100.0   
appliances              Air Conditioners             38725.549240  199.0   
accessories             Gold & Diamond Jewellery     29929.124631  110.0   
sports & fitness        Cardio Equipment             28846.721122  235.0   
appliances              Refrigerators                19792.077584   99.0   
...                                                           ...    ...   
grocery & gourmet foods Coffee, Tea & Beverages        498.725428    8.0   
beauty & health         Beauty & Grooming              481.695149   19.0   
home & kitchen          Sewing & Craft Supplies        438.800707   35.0   
grocery & gourmet foods Snack Foods                    392.350987    9.0   
                        All Grocery & Gourmet Foods    358.693512    9.0   

                                                           max  count  
main_category           sub_category                                   
tv, audio & cameras     Televisions                  1249990.0   1466  
appliances              Air Conditioners              128800.0   1000  
accessories             Gold & Diamond Jewellery      468012.0  31486  
sports & fitness        Cardio Equipment              225250.0    392  
appliances              Refrigerators                 189200.0   3278  
...                                                        ...    ...  
grocery & gourmet foods Coffee, Tea & Beverages         7900.0   2568  
beauty & health         Beauty & Grooming               9200.0   3830  
home & kitchen          Sewing & Craft Supplies        15999.0   2518  
grocery & gourmet foods Snack Foods                     3999.0   2088  
                        All Grocery & Gourmet Foods    13899.0   1908  

[112 rows x 4 columns]

Dengan dua analisis diatas, kita dapat lihat bahwa walaupun 'appliences' dan 'accesories' merupakan `main_category` dengan nilai rata-rata `current_price` paling tinggi. Cukup mengejutkan dalam `sub_category` dipuncaki oleh 'televisions' dimana `main_category`nya merupakan urutan ketujuh dalam `main_category`.

Hal ini bisa kita lihat karena memang 'televisions' memiliki price max yang sangat besar dengan 1249990. 

### Discount

In [26]:
# Membuat kolom discount

df_products['discount'] = df_products['actual_price'] - df_products['current_price']

In [27]:
df_products.groupby('main_category')['discount'].agg(['mean', 'max']).sort_values(by='mean', ascending=False)

mean           max
main_category                                       
home & kitchen           684923.029268  9.899999e+09
sports & fitness           6333.442432  6.108230e+07
appliances                 2668.417563  3.700000e+05
accessories                2520.361441  7.939910e+05
tv, audio & cameras        2298.101095  3.449100e+05
bags & luggage             1730.531382  9.600000e+04
music                      1617.782216  5.244100e+04
men's shoes                1252.831183  2.695600e+04
industrial supplies        1136.832185  4.505000e+04
women's clothing           1110.078051  2.700000e+04
stores                     1036.697163  7.939910e+05
women's shoes               842.148387  4.375900e+04
home, kitchen, pets         824.411765  2.920000e+03
men's clothing              788.685375  1.999600e+04
car & motorbike             780.615048  7.009000e+04
kids' fashion               671.861599  7.939910e+05
toys & baby products        643.387557  3.513500e+04
pet supplies                424.026510  1.353639e+04
beauty & health             401.879788  2.500000e+04
grocery & gourmet foods     162.146636  2.410100e+04

In [28]:
df_products.groupby(['main_category', 'sub_category'])['discount'].agg(['mean', 'sum', 'count']).sort_values(by='mean', ascending=False)

mean  \
main_category           sub_category                                
home & kitchen          Garden & Outdoors            9.017007e+06   
sports & fitness        Running                      6.834569e+04   
tv, audio & cameras     Televisions                  2.238134e+04   
sports & fitness        Cardio Equipment             1.857530e+04   
appliances              Air Conditioners             1.608968e+04   
...                                                           ...   
beauty & health         Health & Personal Care       2.381061e+02   
grocery & gourmet foods Snack Foods                  1.970310e+02   
beauty & health         Value Bazaar                 1.579241e+02   
grocery & gourmet foods All Grocery & Gourmet Foods  1.519794e+02   
                        Coffee, Tea & Beverages      1.413369e+02   

                                                              sum  count  
main_category           sub_category                                      
home & kitchen          Garden & Outdoors            1.980135e+10   2196  
sports & fitness        Running                      1.237057e+08   1810  
tv, audio & cameras     Televisions                  3.281104e+07   1466  
sports & fitness        Cardio Equipment             7.281519e+06    392  
appliances              Air Conditioners             1.608968e+07   1000  
...                                                           ...    ...  
beauty & health         Health & Personal Care       5.228810e+05   2196  
grocery & gourmet foods Snack Foods                  4.114006e+05   2088  
beauty & health         Value Bazaar                 2.084598e+04    132  
grocery & gourmet foods All Grocery & Gourmet Foods  2.899768e+05   1908  
                        Coffee, Tea & Beverages      3.629531e+05   2568  

[112 rows x 3 columns]

Hasil yang ditemuka dari analisis diatas konsisten untuk jumlah rata-rata diskon diuruti paling tinggi dalam `main_category` adalah 'home & kitchen' dan 'sport & fitness' dan juga konsisten dalam `sub_category` yang dipimpin oleh 'Garden & Outdoors' dan 'Running.

'televisions' muncul kembali dalam `sub_category` urutan ketiga.

Kita juga akan melihat bagaimana tidak hanya diskon, tetapi juga persentase diskon dalam suatu produk.

In [29]:
# Membuat kolom 'discount percentage'

df_products['discount_percentage'] = (df_products['actual_price'] - df_products['current_price'])/df_products['actual_price']

In [30]:
df_products.groupby('main_category')['discount_percentage'].agg(['mean', 'max']).sort_values(by='mean', ascending=False)

mean       max
main_category                              
women's clothing         0.565146  0.949703
accessories              0.505547  0.999000
tv, audio & cameras      0.470715  0.991559
home & kitchen           0.458333  1.000000
bags & luggage           0.457284  0.999000
sports & fitness         0.445736  0.999989
car & motorbike          0.433778  0.979098
kids' fashion            0.422090  0.959111
men's clothing           0.412937  0.945541
stores                   0.406195  0.990000
women's shoes            0.393542  0.960096
industrial supplies      0.384635  0.984944
appliances               0.381997  0.996508
music                    0.373426  0.892995
men's shoes              0.365935  0.934147
toys & baby products     0.349717  0.920000
beauty & health          0.316438  0.933489
pet supplies             0.314503  0.932276
grocery & gourmet foods  0.182654  0.945973
home, kitchen, pets      0.172337  0.382826

In [31]:
df_products.groupby(['main_category', 'sub_category'])['discount_percentage'].agg(['mean', 'sum']).sort_values(by='mean', ascending=False)

mean           sum
main_category           sub_category                                       
accessories             Fashion & Silver Jewellery   0.643045  24289.093307
women's clothing        Ethnic Wear                  0.627406  23551.569500
                        Clothing                     0.595726  22676.908288
accessories             Jewellery                    0.575426  21667.077817
women's clothing        Western Wear                 0.558816  21251.765064
...                                                       ...           ...
grocery & gourmet foods All Grocery & Gourmet Foods  0.176166    336.124467
home, kitchen, pets     Refurbished & Open Box       0.172337      5.859442
stores                  Sportswear                   0.148848   2118.100738
grocery & gourmet foods Coffee, Tea & Beverages      0.147204    378.018619
toys & baby products    International Toy Store      0.121031      5.567447

[112 rows x 2 columns]

Hasil yang ditemukan cukup mengejutkan dikarenakan hasil analisis `discount_percentage` dan `discount` cukup berbeda.

Perbedaan itu dapat dilihat dengan bagaiaman `main_category` dipimpin oleh 'women's clothing', 'accesories', dan 'tv, audio, & cameras' dalam sisi `discount_percentage`. Hal ini konsisten dengan analisis `sub_category` dipimpin dengan sub-kategori yang memiliki `main_category` yang konsisten dengan analisis. Dipimpin oleh 'Fashing & Silver Jewellery', 'Ethnic Wear', & 'Clothing'.